clock.go

```diff
@@ -20,95 +16,245 @@ const (
 func AfterFunc(d time.Duration, f func()) *clock.Timer {
-	mutex.Lock()
-	defer mutex.Unlock()
-
-	return Clock.AfterFunc(d, f)
+	return def.AfterFunc(d, f)
 }
 
 // Now returns the current local time.
 func Now() time.Time {
-	mutex.Lock()
-	defer mutex.Unlock()
-
-	return Clock.Now()
+	return def.Now()
 }
 
 // Since returns the time elapsed since t.
 func Since(t time.Time) time.Duration {
-	mutex.Lock()
-	defer mutex.Unlock()
-
-	return Clock.Since(t)
+	return def.Since(t)
 }
 
 // Sleep pauses the current goroutine for at least the duration d.
 func Sleep(d time.Duration) {
-	mutex.Lock()
-	defer mutex.Unlock()
-
-	Clock.Sleep(d)
+	def.Sleep(d)
 }
 
 // Tick is a convenience wrapper for NewTicker providing access to the ticking channel only.
 func Tick(d time.Duration) <-chan time.Time {
-	mutex.Lock()
-	defer mutex.Unlock()
-
-	return Clock.Tick(d)
+	return def.Tick(d)
 }
 
 // Ticker returns a new Ticker containing a channel that will send the
 // time with a period specified by the duration argument.
 func Ticker(d time.Duration) *clock.Ticker {
-	mutex.Lock()
-	defer mutex.Unlock()
-
-	return Clock.Ticker(d)
+	return def.Ticker(d)
 }
 
 // Timer creates a new Timer that will send the current time on its channel after at least duration d.
 func Timer(d time.Duration) *clock.Timer {
-	mutex.Lock()
-	defer mutex.Unlock()
+	return def.Timer(d)
+}
 
-	return Clock.Timer(d)
+func Until(t time.Time) time.Duration {
+	return def.Until(t)
 }
 
-// Mock replaces the Clock with a mock frozen at the given time and returns it.
-func Mock(now time.Time) *clock.Mock {
-	mutex.Lock()
-	defer mutex.Unlock()
+func WithDeadline(parent context.Context, d time.Time) (context.Context, context.CancelFunc) {
+	return def.WithDeadline(parent, d)
+}
+
+func WithTimeout(parent context.Context, t time.Duration) (context.Context, context.CancelFunc) {
+	return def.WithTimeout(parent, t)
+}
+
+// Current returns current instance of clock
+func Current() clock.Clock {
+	return def.get()
+}
 
+// Mock sets the Clock with a mock frozen at the given time and returns it.
+func Mock(now time.Time) *clock.Mock {
 	mock := clock.NewMock()
 	mock.Set(now)
 
-	Clock = mock
+	def.set(mock)
 
 	return mock
 }
 
+// Set sets the clock.
+func Set(clk clock.Clock) {
+	def.set(clk)
+}
+
 // Restore replaces the Clock with the real clock.
 func Restore() {
-	mutex.Lock()
-	defer mutex.Unlock()
+	def.restore()
+}
+
+// UseLock adds locking mechanism back on clock.
+func UseLock() {
+	def.setLock()
+}
+
+// NoLock removes locking mechanism usage on clock.
+func NoLock() {
+	def.setNoLock()
+}
+
+type mutexWrap struct {
+	lock     sync.Mutex
+	disabled bool
+}
+
+func (mw *mutexWrap) Lock() {
+	if !mw.disabled {
+		mw.lock.Lock()
+	}
+}
+
+func (mw *mutexWrap) Unlock() {
+	if !mw.disabled {
+		mw.lock.Unlock()
+	}
+}
+
+func (mw *mutexWrap) Enable() {
+	mw.lock.Lock()
+	defer mw.lock.Unlock()
+
+	mw.disabled = false
+}
+
+func (mw *mutexWrap) Disable() {
+	mw.lock.Lock()
```

Comment:
Could this cause issue? Lock is done when disabled = false, Unlock is attempted when disabled = true